src/ray/core_worker/core_worker.cc

```diff
@@ -560,25 +585,45 @@ bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
     // Register a callback to handle actor notifications.
     auto actor_notification_callback = [this](const ActorID &actor_id,
                                               const gcs::ActorTableData &actor_data) {
-      if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
-        auto it = actor_handles_.find(actor_id);
-        RAY_CHECK(it != actor_handles_.end());
-        if (it->second->IsDirectCallActor()) {
-          // We have to reset the actor handle since the next instance of the
-          // actor will not have the last sequence number that we sent.
-          // TODO: Remove the check for direct calls. We do not reset for the
-          // raylet codepath because it tries to replay all tasks since the
-          // last actor checkpoint.
-          it->second->Reset();
+      auto it = actor_handles_.find(actor_id);
+      RAY_CHECK(it != actor_handles_.end());
+      if (actor_data.state() == gcs::ActorTableData::ALIVE) {
+        it->second->Connect(actor_data, client_call_manager_);
+        direct_actor_submitter_->SendPendingTasks(actor_id, it->second->RpcClient());
+      } else {
+        // We have to reset the actor handle since the next instance of the
+        // actor will not have the last sequence number that we sent.
+        // TODO: Remove the flag for direct calls. We do not reset for the
+        // raylet codepath because it tries to replay all tasks since the last
+        // actor checkpoint.
+        it->second->Reset(actor_data,
+                          /*reset_task_counter=*/it->second->IsDirectCallActor());
+        direct_actor_submitter_->FailPendingTasks(actor_id);
+
+        if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
+          // If we are the actor's creator, restart it.
+          auto child_it = children_actors_.find(actor_id);
+          if (child_it != children_actors_.end()) {
+            child_it->second.num_lifetimes++;
+            // Restart the actor.
+            if (!raylet_client_
+                     ->SubmitTask(child_it->second.actor_creation_spec,
+                                  child_it->second.num_lifetimes)
+                     .ok()) {
+              // TODO: Mark the actor as DEAD in the GCS. This is currently
+              // hard to do because the GCS first expects an ALIVE entry in the
+              // actor's log before we add the DEAD entry.
+              RAY_LOG(WARNING) << "Failed to restart actor " << actor_id;
```

Comment:
Can we log loudly when-ever we restart an actor (ERROR level seems suitable). This should be a highly visible event to the user.